test/python/primitives/test_backend_estimator_dev.py

```diff
@@ -0,0 +1,924 @@
+            - Classical bits remain in order in measurements through transpilation
+        """
+        # Input and measured circuits
+        num_qubits = len(layout_intlist)
+        input_circuit = QuantumCircuit(num_qubits)
+        measured_circuit = input_circuit.copy()
+        measured_circuit.measure_all()
+        # Transpiled circuit (only changes layout and num of qubits)
+        layout_dict = dict.fromkeys(range(target_qubits))
+        layout_dict.update(dict(zip(layout_intlist, input_circuit.qubits)))
+        applied_layout = Layout(layout_dict)
+        passes = [SetLayout(layout=applied_layout), ApplyLayout()]
+        pass_manager = PassManager(passes=passes)
+        transpiled_circuit = pass_manager.run(measured_circuit)  # TODO: skip_transpilation
+        # Test patching terra's transpile call
+        backend = Mock(Backend)
+        estimator = BackendEstimator(backend)
+        estimator._transpile_options = MagicMock(Options)
+        with patch("qiskit.primitives.backend_estimator_dev.transpile", spec=True) as mock:
+            mock.return_value = transpiled_circuit
+            output_circuit = estimator._transpile(input_circuit)
+        mock.assert_called_once()
+        (call_circuit, call_backend), call_kwargs = mock.call_args
+        self.assertEqual(call_circuit, measured_circuit)
+        self.assertIs(call_backend, backend)
+        self.assertEqual(call_kwargs, estimator._transpile_options.__dict__)
+        self.assertEqual(output_circuit, transpiled_circuit)
+        self.assertIsInstance(output_circuit, QuantumCircuit)
+        inferred_layout = output_circuit.metadata.get("final_layout")
+        self.assertEqual(inferred_layout, applied_layout)
+        self.assertIsInstance(inferred_layout, Layout)
+
+    def test_run_bound_pass_manager(self):
+        """Test bound pass manager runs."""
+        backend = Mock(Backend)
+        estimator = BackendEstimator(backend)
+        # Invalid input
+        self.assertRaises(TypeError, estimator._run_bound_pass_manager, "circuit")
+        # No pass manager
+        circuit = Mock(QuantumCircuit)
+        self.assertIs(circuit, estimator._run_bound_pass_manager(circuit))
+        # Pass manager runs
+        mock_circuit = Mock(QuantumCircuit)
+        estimator._bound_pass_manager = Mock(PassManager)
+        estimator._bound_pass_manager.run.return_value = mock_circuit
+        self.assertIs(mock_circuit, estimator._run_bound_pass_manager(circuit))
+        estimator._bound_pass_manager.run.assert_called_once_with(circuit)
+
+
+@ddt
+class TestMeasurement(TestCase):
+    """Test measurement logic."""
+
+    def test_observable_decomposer(self):
+        """Test observable decomposer property."""
+        estimator = BackendEstimator(Mock(Backend))
+        self.assertTrue(estimator.abelian_grouping)
+        self.assertIsInstance(estimator._observable_decomposer, AbelianDecomposer)
```

Comment:
Testing private components like this is generally not a good idea.  They aren't part of the public interface, and writing hyper-explicit tests like this means that the structure of the underlying code cannot change without the tests also changing, which very easily masks bugs.  There's a good amount of confidence given by a PR that modifies code, and _all_ the tests easily pass, and it's very concerning when a PR _needs_ to modify the tests to get them to pass, unless the test is clearly buggy.

Reply:
Interesting 🤔 this is my way of making sure that the behavior for `abelian_grouping` stays consistent. True, it is a private property, but merely a reflection of a user configurable attribute. 

I understand what you are saying and think that it may apply to other places, but I am not so sure here. If we change a what `abelian_grouping` means, shouldn't the tests fail? How would yo handle this otherwise?

Reply:
> Presumably the user setting abelian_grouping=True should cause some changes to the output they see from public methods?

Negative... it only changes the way things are done internally. This is inherited from the original code.

Reply:
Since transpiled_circuits are exposed, it can be tested from the length of transpiled QuantumCircuit. (or preprocessed circuits)